slides/03_workflow.qmd

```diff
@@ -1,10 +1,591 @@
+        mean1 = x$mean1, mean2 = x$mean2, sd1 = x$sd1, sd2 = x$sd2)
+    
+    print(list(
+        args = format(args), 
+        data = dplyr::tibble(x$data)
+    ), ...)
+}
+x
+```
+
+## Roxygen
+
+```{r, echo = TRUE, eval = FALSE}
+#' @title
+#' Print Simulation Result
+#'
+#' @description
+#' Generic function to print a \code{SimulationResult} object.
+#'
+#' @param x a \code{SimulationResult} object to print.
+#' @param ... further arguments passed to or from other methods.
+#' 
+#' @examples
+#' x <- getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, 
+#'      mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
+#' print(x)
+#'
+#' @export
+```
+
+## Output
+
+```{r, echo = FALSE}
+#| style: font-size:0.8em!important;
+print.SimulationResult <- function(x, ...) {
+    args <- list(n1 = x$n1, n2 = x$n2, 
+        mean1 = x$mean1, mean2 = x$mean2, sd1 = x$sd1, sd2 = x$sd2)
+    
+    print(list(
+        args = format(args), 
+        data = dplyr::tibble(x$data)
+    ), ...)
+}
+x
+```
+:::
+
+# Exercise
+
+![[Photo CC0 by Pixabay on
+pexels.com]{.copyright}](thumbnails/exercise.jpg){fig-align="center"}
+
+## Preparation
+
+1.  Download the unfinished R package [simulatr](download/simulatr.zip)
+2.  Extract the package zip file
+3.  Open the project with RStudio
+4.  Complete the tasks below
+
+## Tasks
+
+::: panel-tabset
+## Task 1
+
+Add assertions to improve the usability and user experience
+
+::: callout-note
+## Tip on assertions
+
+Use the package
+[checkmate](https://cran.r-project.org/package=checkmate "checkmate: Fast and Versatile Argument Checks")
+to validate input arguments.
+:::
+
+Example:
+
+```{r, echo = TRUE, eval = FALSE}
+playWithAssertions <- function(n1) {
+  checkmate::assertInt(n1, lower = 1)
+}
+playWithAssertions(-1)
+```
+
+[Error in playWithAssertions(-1) :]{.error} [Assertion on 'n1' failed: Element 1
+is not \>= 1.]{.error}
+
+## Task 2
+
+Add three additional results:
+
+1.  n total,
+2.  creation time, and
+3.  allocation ratio
+
+::: callout-note
+## Tip on creation time
+
+`Sys.time()`, `format(Sys.time(), '%B %d, %Y')`, `Sys.Date()`
+:::
+
+## Task 3
+
+Add an additional result: `t.test` result
+
+## Task 4
+
+Add an optional *alternative* argument and pass it through `t.test`:\
+
+```{r, echo = TRUE}
+#| code-line-numbers: false
+alternative = c("two.sided", "less", "greater")
+```
+
+## Task 5
+
+Implement the generic functions `print` and `plot`.
+
+::: callout-note
+## Tip on print
+
+Use the plot example function from above and extend it.
+:::
+
+::: callout-note
+## Tip on plot
+
+Use R base
+[plot](https://bookdown.org/rdpeng/exdata/the-base-plotting-system-1.html) or
+[ggplot2](https://cran.r-project.org/package=ggplot2) to create a grouped
+boxplot of the fake data.
+:::
+
+## Task 6
+
+Implement the generic functions
+
+-   `summary`,
+-   `kable`, and
+-   `cat`.
+
+## Task 7
+
+Document your functions with
+[Roxygen2](https://cran.r-project.org/package=roxygen2)
+:::
```

Comment:
for these exercises - I guess you are going to show this yourself in the class? because we don't have time so far allocated to this as far as I see from the schedule

Reply:
Time for excercises is planned for this session, see
10:45 - 12:00 | Workflow for creating R packages + practical